### ISSUE TITLE
kill-this-buffer stops working on OS X

### ISSUE THREAD

**Issue by jupl** [syl20bnr/spacemacs#4929]
#### Description
When using OS X with `dotspacemacs-persistent-server` enabled, buffers can be closed with the keystrokes `SPC b d`. However, when closing window and opening a new one after keystroke no longer works. `SPC b k ENTER` does work. (extra keystroke required though) This is happening in both master and develop.

#### Reproduction guide
- Start Emacs
- `SPC f e d` to open a buffer
- `SPC b d` to close the buffer
- `:q` to close window
- Click on app in dock (app should still be running) to open a new window
- `SPC f e d` to open a buffer
- `SPC b d` to close the buffer

*Observed behaviour:*
Only the first use of `SPC b d` in the reproduction steps closes the buffer. The second one does not work. Messages buffer does not show any error messages.

*Expected behaviour:*
Both instances of `SPC b d` should close the buffer as expected.

**Comment by robbyoconnor** [comment]
@jupl Thanks, but follow this template: https://github.com/syl20bnr/spacemacs/wiki/Debugging#how-to-make-a-great-bug-report note that if you're in develop `SPC h I` will open an issue prefilled with all that template.

**Comment by sooheon** [comment]
FWIW, I can't seem to reproduce this after `:q` or `spc q z`

**Comment by jupl** [comment]
@robbyoconnor Done.

**Comment by kccai** [comment]
I cannot reproduce the issue. But `SPC b d` does not work occasionally (don't know why). I have to restart emacs every time this happens.

**Comment by mineo** [comment]
`SPC b d` stops working for me as well after some time. `menu-bar-menu-frame-live-and-visible-p` is always nil at that point, which means that `kill-this-buffer` simply doesn't do anything. Restarting Emacs fixes that, `SPC q z` and opening a new window makes `menu-bar-...` be nil again.

**Comment by kccai** [comment]
Yes, I also have this problem on my Arch Linux machine. I do not think this is a OS X specific issue.

**Comment by levindu** [comment]
This also happens in Linux.
But if  SPC b d not works, just run 'M-x menu-bar-mode' twice (toggle menu bar), and it work again. No need to restart emacs.

**Comment by bluesd** [comment]
Have this problem too (Linux and Windows), seems to be related with kill-this-buffer when using emacsclient. 

**You can reproduce it just doing:**
* Open `emacs`
* `SPC q z`
* Open `emacs -c`
* `SPC b d`. -> Does nothing.

As long as you don't close the first `emacs` window, `SPC b d` behaves as expected.

Maybe remapping `SPC b z` from `kill-this-buffer` to `evil-delete-buffer` could be a workaround? since `evil-delete-buffer`  works even in this situation. (Same for `SPC b k`, etc).

/// By the way, `menu-bar.el` also mentions this bug:
```
   ;; Don't do anything when `menu-frame' is not alive or visible
   ;; (Bug#8184).
```

**Comment by StreakyCobra** [comment]
> Maybe remapping SPC b d from kill-this-buffer to evil-delete-buffer could be a workaround? since evil-delete-buffer works even in this situation. (Same for SPC b k, etc).

Better askin' the specialist. @bmag :point_up:

**Comment by bmag** [comment]
I only did a shallow research. It seems that `menu-updating-frame` points to an invisible frame, but I guess its value should be nil. Perhaps the persistent-server feature works by creating an invisible frame, and that's what confuses `menu-updating-frame`? Does this bug happen only with the persistent-server enabled?

Anyway, it looks like `kill-this-buffer` is intended to be used only from the menu-bar. We probably should use a different function. I don't think `evil-delete-buffer` is good, because it also closes the window (if there's more than 1 window). It shouldn't be too hard to find a suitable function, or write one, if somebody is up to it.

**Comment by bluesd** [comment]
At least for me happens only with the `dotspacemacs-persistent-server` enabled AND only if you close (`SPC q z`) the first `emacs` you invoke. After that, the current and future `emacsclient -c` never return when closing the frame, leaving the terminal "Waiting for emacs..." without resuming to shell.

Also, in this "state", if you attempt to `SPC q Q`, the first time doesn't work at all and you get the following legend: "Attempt to delete the sole visible or iconified frame". (I think this one's only appearing in Linux).

**Comment by toupeira** [comment]
This works nicely for me, and still asks for confirmation in modified buffers:

```elisp
(spacemacs/set-leader-keys
  "bd" (lambda () (interactive) (kill-buffer (current-buffer))))
```

**Comment by bmag** [comment]
@toupeira that works for me too. Do you want to submit a pull request for this change? If not, let me know and I'll submit one.

**Comment by toupeira** [comment]
@bmag why not, I submitted one :)

**Comment by StreakyCobra** [comment]
Is this issue fixed now that #6225 has been merged? If yes I'll mark it as "Fixed in develop"

**Comment by toupeira** [comment]
@StreakyCobra yes it works fine for me in develop.

**Comment by d12frosted** [comment]
Fixed with released of Spacemacs v0.200. Let us know if you still have any problems with this issue.
